kafka/codec.py

```diff
@@ -299,3 +309,18 @@ def lz4_decode_old_kafka(payload):
         payload[header_size:]
     ])
     return lz4_decode(munged_payload)
+
+
+def zstd_encode(payload):
+    if not zstd:
+        raise NotImplementedError("Zstd codec is not available")
+    return zstd.ZstdCompressor().compress(payload)
+
+
+def zstd_decode(payload):
+    if not zstd:
+        raise NotImplementedError("Zstd codec is not available")
+    try:
+        return zstd.ZstdDecompressor().decompress(payload)
+    except zstd.ZstdError:
+        return zstd.ZstdDecompressor().decompress(payload, max_output_size=ZSTD_MAX_OUTPUT_SIZE)
```

Comment:
Reading the docs on [`ZstdError`](https://github.com/indygreg/python-zstandard#simple-api-1) it looks like this can still cause problems if the output is larger than `ZSTD_MAX_OUTPUT_SIZE`... 

It looks like the value of `ZSTD_MAX_OUTPUT_SIZE` is currently set to `1024 ** 3`... Why did you pick that particular value?

I realize this is a fallback, with the expectation that in the normal case the zstd header will contain the frame size, so I think it's okay to set a fallback that works for the majority of cases but may not work in all cases...

Reply:
I actually made this fail using the java console producer, which does not properly complete the message frame with size information. I used this line https://github.com/facebook/zstd/blob/dev/doc/educational_decoder/harness.c#L90 as inspiration, and while i do realise that value does not come into play in any shape inside the core algorithm, it was better than nothing

Reply:
though in retrospect this will probably end up causing a lot of OOM errors in a traffic intensive scenario, since it tries to allocate 1G or memory per message with frame information missing. Ideally this would be taken from the broker configuration `"message.max.bytes"`, but i see no way of passing that in

Reply:
Agreed that `message.max.bytes` would be ideal... for now, why don't we just set it to 1 MB, since that's the default for `message.max.bytes`?

I expect that won't be perfect, but it'll be a start and let us land this and then see what further suggestions/feedback arise from others... some of this ultimately comes back to it'd be better if we could let compression be a passed in class so that users can tweak their own settings external to the client. But that's a much bigger change far beyond the scope of this PR.

Reply:
done